src/site/content/en/blog/streams/index.md

```diff
@@ -0,0 +1,966 @@
+  // `value` - Some data. Always `undefined` when `done` is `true`.
+  if (done) {
+    return console.log(`Stream complete.\n${result}`);
+  }
+  result += value;
+  console.log(`Read ${result.length} characters so far`);
+  console.log(`Most recently read chunk: ${value}`);
+  // Read some more, and call this function again.
+  return readStream();
+};
+readStream();
+```
+
+The next (a bit contrived) code sample shows how you could implement a "shouting" service worker
+that uppercases all text by consuming the returned `fetch()` response promise
+[as a stream](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API/Using_readable_streams#consuming_a_fetch_as_a_stream)
+and uppercasing chunk by chunk. The advantage of this approach is that you do not need to wait for
+the whole document to be downloaded, which can make a huge difference when dealing with large files.
+
+```js
+// Install immediately without waiting.
+self.addEventListener('install', (event) => {
+  self.skipWaiting();
+});
+
+// Claim all clients immediately.
+self.addEventListener('activate', (event) => {
+  clients.claim();
+});
+
+self.addEventListener('fetch', (event) => {
+  // If the destination of the request is anything else
+  // than a document, return.
+  if (event.request.destination !== 'document') {
+    return;
+  }
+  event.respondWith(
+    (async () => {
+      // Create a new stream that implements the UPPERCASING.
+      const result = await shoutFetch(event.request.url);
+      // Return the newly created stream.
+      return new Response(result.stream, { headers: result.headers });
+    })(),
+  );
+});
+
+const shoutFetch = async (url) => {
+  // Chunks arrive encoded, so need to decode them.
+  const textDecoder = new TextDecoder('utf-8', { stream: true });
+  // After uppercasing, need to encode chunks again.
+  const textEncoder = new TextEncoder();
+  // Fetch the initial document and get access to the
+  // reader of the readable stream.
+  const response = await fetch(url);
+  const readableStream = response.body;
+  const reader = readableStream.getReader();
+  // Create a new stream.
+  const responseStream = new ReadableStream({
+    // Called whenever the internal queue of chunks is not full.
+    async pull(controller) {
+      // Read the next chunk.
+      let result = await reader.read();
+      // Close the stream when done.
+      if (result.done) {
+        return controller.close();
+      }
+      // Uppercase each chunk (after decoding it).
+      const upperCaseValue = textDecoder.decode(result.value).toUpperCase();
+      // Enqueue the uppercased value (afer re-encoding it).
+      controller.enqueue(textEncoder.encode(upperCaseValue));
+      return result;
+    },
+
+    // Called when the reader cancels.
+    cancel() {
+      reader.cancel();
+    },
+  });
+  // Return the new stream with the headers of the original response.
+  return await { stream: responseStream, headers: response.headers };
+};
+```
```

Comment:
```suggestion
```

{% Aside %}
  This example would be much simpler as a transform stream,
  which I will introduce further down in the article. Stay tuned!
{% endAside %}
```